File: 04web_iot/04web_iot.py
from flask import Flask, request, redirect
import RPi.GPIO as GPIO
from flask_socketio import SocketIO, send
import Adafruit_DHT, json, time

app = Flask(__name__)
app.config['SECRET_KEY'] = 'secret'
socketio = SocketIO(app,async_mode='threading')

pin_led = 18
pin_btn = 23
pin_dht11 = 24

GPIO.setmode(GPIO.BCM)
GPIO.setup(pin_led, GPIO.OUT)
GPIO.setup(pin_btn, GPIO.IN)
sensor = Adafruit_DHT.DHT11

@app.route('/')
def main():
    return redirect('/static/web_iot.html')


@app.route('/operate/<cmd>')
def op(cmd):
    val = request.values['val']   
    if cmd == "led":
        val = request.values['val']   
        print '/operate/', cmd, val 
        if val == 'on':
            GPIO.output(pin_led, True)
            print pin_led, 'on'
        elif val == 'off':
            GPIO.output(pin_led, False)
        return 'OK'


@app.route('/monitor')
def monitoring():
    try:
        print '/monitor'
        humidity, temperature = Adafruit_DHT.read_retry(sensor, pin_dht11)
        obj = {'humi' : humidity, 'temp' : temperature}
        print obj
        return json.dumps(obj)
    except Exception as e:
        print 'err', e


@socketio.on('connect')
def connect():
    send({'data': 'welcome'})
    
def my_callback(channel):
    print 'Edge detected on channel %s state %s'% (channel, GPIO.input(channel)) 
    val =GPIO.input(channel)
    if val == 1:
        socketio.send( {'data': '1' })
    else:
        socketio.send({'data': '0'})
    print 'btn pressed', 0

if __name__ == '__main__':
    try:
        GPIO.add_event_detect(pin_btn, GPIO.BOTH, callback=my_callback)
        socketio.run(app, host='0.0.0.0')
    finally:
        print 'cleaning up'
        GPIO.cleanup()

File: 04web_iot/04web_iot.js
var http = require('http');
var express = require('express');
var socketio = require('socket.io');
var onoff =require('onoff');
var sensor = require('node-dht-sensor');

var app = express();
var server = http.createServer(app);
var io = socketio(server);

var pin_led = new onoff.Gpio(18, 'out');
var sw_pin = new onoff.Gpio(23, 'in', 'both');
var dht_type = 11;
var pin_dht = 24

sw_pin.watch(function(err, value){
	console.log('btn: '+ value);
	io.emit('message', {data: value});
});


app.use(express.static(__dirname + '/static'));
app.get('/', function(req, res){
	res.redirect('/web_iot.html');
});

app.get('/operate/led', function(req, res){
	console.log(req.url, req.query);
	var val = req.param('val');
	if(val == 'on'){
		pin_led.writeSync(1);
		res.send('OK');
	}else if(val ==='off'){
		pin_led.writeSync(0);
		res.send('OK');
	}else{
		res.send('Fail');
	}
});

app.get('/monitor', function(req, res){
	console.log(req.url, req.query);
	sensor.read(dht_type, pin_dht, function(err, temperature, humidity){
		if(!err){
			var obj = {'humi' : humidity, 'temp' : temperature};
		}else{
			var obj = {'humi' : 'err', 'temp' : 'err'};
		}
		console.log(JSON.stringify(obj));
		res.send(obj);
	});
});

io.on('connection', function(socket){
	console.log('connected' + socket.id);
	 socket.emit('message',{'data': 'welcome'});
});

server.listen(5000, function(){
	console.log('server running on 5000.');
});